File: quarto-semestre/AEDS 2/tps_resolvidos/tp01/ex15/Arquivo.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    FILE *arquivo;
    int n;

    arquivo = fopen("valores.txt", "wb");
    if (arquivo == NULL) {
        perror("Erro ao abrir o arquivo");
        return 1;
    }

    scanf("%d", &n);
    getchar();

    for (int i = 0; i < n; i++) {
        char valorStr[20];
        fgets(valorStr, 20, stdin);

        // Substitui vírgula por ponto
        for (int j = 0; valorStr[j] != '\0'; j++) {
            if (valorStr[j] == ',') {
                valorStr[j] = '.';
            }
        }

        double valor = atof(valorStr); 
        fwrite(&valor, sizeof(double), 1, arquivo); 
    }

    fclose(arquivo);

    arquivo = fopen("valores.txt", "rb");

    // Move o ponteiro para o final do arquivo
    fseek(arquivo, 0, SEEK_END);
    long posicao = ftell(arquivo);

    // Leitura do arquivo de trás para frente
    for (int i = 0; i < n; i++) {
        posicao -= sizeof(double);
        fseek(arquivo, posicao, SEEK_SET);

        double valor;
        fread(&valor, sizeof(double), 1, arquivo);

        // Exibe os valores sem zeros desnecessários à direita
        if (valor == (int)valor) {
            printf("%.0f\n", valor);  // Exibe como inteiro
        } else {
            
            printf("%.15g\n", valor); // Usa %.15g para remover zeros à direita
        }
    }

    fclose(arquivo);

    return 0;
}

File: quarto-semestre/AEDS 2/tps_resolvidos/tp01/ex14/Arquivo.java
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class Arquivo {

    public static void main(String[] args) {
        try {
           
            RandomAccessFile arquivo = new RandomAccessFile("valores.txt", "rw");

            
            Scanner scanner = new Scanner(System.in);
            int n = scanner.nextInt();  
            scanner.nextLine();

            // Leitura dos números reais e gravação no arquivo
            for (int i = 0; i < n; i++) {
               
                String valorStr = scanner.nextLine().replace(",", "."); // Substitui vírgula por ponto porque tava dando erro no scanner
                double valor = Double.parseDouble(valorStr);  
                arquivo.writeDouble(valor); 
            }

            arquivo.close();
            scanner.close();

            arquivo = new RandomAccessFile("valores.txt", "r");

            // Mover o ponteiro para o final do arquivo
            long posicao = arquivo.length();


            // Leitura do arquivo de trás para frente
            for (int i = 0; i < n; i++) {

                // Mover o ponteiro para o valor anterior
                posicao -= 8;  // Cada valor double ocupa 8 bytes
                arquivo.seek(posicao);

    
                double valor = arquivo.readDouble();

                // Se a parte decimal for 0, exibe o valor como inteiro
                if (valor == (int) valor) {
                    System.out.println((int) valor); 
                } else {
                    System.out.println(valor); 
                }
            }

            arquivo.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
